packages/react/src/components/cookieConsent/CookieConsentContext.tsx

```diff
@@ -0,0 +1,283 @@
+
+import createConsentController, { ConsentController, ConsentList, ConsentObject } from './cookieConsentController';
+import { ContentSource, createContent } from './content.builder';
+
+export type Description = {
+  title: string;
+  text: string;
+};
+
+export type TableData = {
+  name: string;
+  hostName: string;
+  description: string;
+  expiration: string;
+};
+
+export type CookieData = TableData & {
+  id: string;
+};
+
+export type CookieGroup = Description & {
+  expandAriaLabel: string;
+  checkboxAriaDescription?: string;
+  cookies: CookieData[];
+};
+
+export type UiTexts = {
+  showSettings: string;
+  hideSettings: string;
+  approveAllConsents: string;
+  approveRequiredAndSelectedConsents: string;
+  approveOnlyRequiredConsents: string;
+  settingsSaved: string;
+  readMore: string;
+};
+
+export type SectionTexts = {
+  main: Description;
+  details: Description;
+};
+
+export type Category = Description & {
+  checkboxAriaDescription?: string;
+  groups: CookieGroup[];
+};
+
+export type SupportedLanguage = 'fi' | 'sv' | 'en';
+
+export type Content = {
+  requiredCookies?: Category;
+  optionalCookies?: Category;
+  texts: {
+    sections: SectionTexts;
+    ui: UiTexts;
+    tableHeadings: TableData;
+  };
+  language: {
+    languageOptions: { code: SupportedLanguage; label: string }[];
+    current: SupportedLanguage;
+    languageSelectorAriaLabel: string;
+    onLanguageChange: (newLanguage: string) => void;
+  };
+  onAllConsentsGiven?: (consents: ConsentObject) => void;
+  onConsentsParsed?: (consents: ConsentObject, hasUserHandledAllConsents: boolean) => void;
+  focusTargetSelector?: string;
+};
+
+export type CookieConsentContextType = {
+  hasUserHandledAllConsents: () => boolean;
+  content: Content;
+  onAction: CookieConsentActionListener;
+  countApprovedOptional: () => number;
+  areGroupConsentsApproved: (consents: CookieData[]) => boolean;
+};
+
+type CookieConsentContextProps = {
+  cookieDomain?: string;
+  children: React.ReactNode | React.ReactNode[] | null;
+  contentSource: ContentSource;
+};
+
+type CookieConsentAction =
+  | 'approveAll'
+  | 'approveRequired'
+  | 'changeCookieGroupConsents'
+  | 'approveOptional'
+  | 'unapproveOptional'
+  | 'approveSelectedAndRequired';
+
+type CookieConsentActionListener = (action: CookieConsentAction, consents?: string[], value?: boolean) => void;
+
+export const CookieConsentContext = createContext<CookieConsentContextType>({
+  hasUserHandledAllConsents: () => false,
+  content: {} as Content,
+  onAction: () => undefined,
+  countApprovedOptional: () => 0,
+  areGroupConsentsApproved: () => false,
+});
+
+export const getConsentsFromCookieGroup = (groups: CookieGroup[]): ConsentList => {
+  return groups.reduce((ids, currentGroup) => {
+    currentGroup.cookies.forEach((consentData) => {
+      ids.push(consentData.id);
+    });
+    return ids;
+  }, []);
+};
+
+const forceFocusToElement = (elementSelector: string): void => {
```

Comment:
Name has "force" in it, because tabIndex is forced, if first focus() fails. It fails, if element is not a focusable element. Setting tabindex fixes this. This way the focus always succeeds.

What would exactly be gained by making this function strictly usable only with one selector/element? Only the argument would be cleared and element selector would be fetched inside the function.

Reply:
The `tabindex` should only matter on whether the element is reachable via sequential keyboard navigation. You should be always able to focus element through javascript: https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/tabindex

> What would exactly be gained by making this function strictly usable only with one selector/element? Only the argument would be cleared and element selector would be fetched inside the function.

It would reveal more information to the reader of the code, without having to look for the implementation. Currently from the api user thinks that this is a general `foruceFocusToElement` function, which implies that there are multiple use cases for it. We could then rename it to be more revealing of the intented usage. Just a minor readability improvement imo, but nothing too fancy so if you think it's better this way, we can leave it as it is.

Reply:
"You should be always able to focus element through javascript". What do you mean?

`document.querySelector("div").focus()`does nothing, if div has no tabindex set.

We can remove the failsafe and just silently fail, if the implementing developer fails to provide a selector for an element which can be focused. Usually a heading is focused after the modal is closed, and <h1> are not focusable by default.

Reply:
> "You should be always able to focus element through javascript". What do you mean?

You're right, it does seem require the tabindex 👍